[red-knot] Add `__init__` arguments check when doing `try_call` on a class literal (#16512)
astral-sh/ruff: crates/red_knot_python_semantic/resources/mdtest/call/constructor.md

constructor from it.

```py
class Base:
    def __init__(self, x: int): ...

class Foo(Base): ...

reveal_type(Foo(1))  # revealed: Foo

# error: [missing-argument] "No argument provided for required parameter `x` of bound method `__init__`"
reveal_type(Foo())  # revealed: Foo
# error: [too-many-positional-arguments] "Too many positional arguments to bound method `__init__`: expected 1, got 2"
reveal_type(Foo(1, 2))  # revealed: Foo
```

## Conditional `__init__`

```py
def _(flag: bool) -> None:
    class Foo:
        if flag:
            def __init__(self, x: int): ...
        else:
            def __init__(self, x: int, y: int = 1): ...

    reveal_type(Foo(1))  # revealed: Foo
    # error: [invalid-argument-type] "Object of type `Literal["1"]` cannot be assigned to parameter 2 (`x`) of bound method `__init__`; expected type `int`"
    reveal_type(Foo("1"))  # revealed: Foo
    # error: [missing-argument] "No argument provided for required parameter `x` of bound method `__init__`"
    reveal_type(Foo())  # revealed: Foo
```

## A descriptor in place of `__init__`

```py
from typing import Callable

class SomeCallable:
    def __call__(self, x: int) -> str:

>         return "a"

Author: Ha. I checked that descriptor works before even starting working on pr. But I've used other code, this one was copied from another test. Hence the "error" with a non-None return type. 

I suggest we keep the test and later emit error diagnostics for non None return even though it works in runtime. I will just add a comment that knot is stricter / deviates from runtime. If we ever see such code it is most probably an error

# Constructor

## No init or new

Every class has `object` in it's MRO, so if no `__init__` method is provided, we fall back to
`object.__init__`, which can only be called with zero arguments:

```py
class Foo: ...

reveal_type(Foo())  # revealed: Foo

# error: [too-many-positional-arguments] "Too many positional arguments to bound method `__init__`: expected 0, got 1"
reveal_type(Foo(1))  # revealed: Foo
```

## `__new__` present on the class itself

If the class has an `__new__` method, we can infer the signature of the constructor from it.

```py
class Foo:
    def __new__(cls, x: int) -> "Foo": ...

reveal_type(Foo(1))  # revealed: Foo

# error: [missing-argument] "No argument provided for required parameter `x` of bound method `__new__`"
reveal_type(Foo())  # revealed: Foo
# error: [too-many-positional-arguments] "Too many positional arguments to bound method `__new__`: expected 1, got 2"

> reveal_type(Foo(1, 2))  # revealed: Foo

Author: thanks for digging into CPYthon source code, that’s definitely easier than reverse-engineering the behavior. It was well past 1am yesterday when I was testing this, and I only managed to muster strength to run some code tests ;-)

anyways, as mypy_prymer results shows, as well mdtest, it doesn’t look like descriptor protocol behaves here as expected. It seems to select up `type.__new__` for one reason or another… I’ll continue tracing why.

> reveal_type(Foo(1, 2))  # revealed: Foo

Reviewer: > It seems to select up `type.__new__` for one reason or another… I’ll continue tracing why.

One thing that might be required is to special case `type.__new__` for the moment. We do not understand overloads yet, so if we generate a `@Todo` type for `type.__new__`, we may end up thinking that it is a data descriptor (because a dynamic type has every attribute, and so it also has a `__set__` attribute), and call that instead.

class Foo(Base): ...

reveal_type(Foo(1))  # revealed: Foo

# error: [missing-argument] "No argument provided for required parameter `x` of function `__new__`"
reveal_type(Foo())  # revealed: Foo
# error: [too-many-positional-arguments] "Too many positional arguments to function `__new__`: expected 1, got 2"
reveal_type(Foo(1, 2))  # revealed: Foo
```

## Conditional `__new__`

```py
def _(flag: bool) -> None:
    class Foo:
        if flag:
            def __new__(cls, x: int): ...
        else:
            def __new__(cls, x: int, y: int = 1): ...

    reveal_type(Foo(1))  # revealed: Foo
    # error: [invalid-argument-type] "Object of type `Literal["1"]` cannot be assigned to parameter 2 (`x`) of function `__new__`; expected type `int`"
    reveal_type(Foo("1"))  # revealed: Foo
    # error: [missing-argument] "No argument provided for required parameter `x` of function `__new__`"
    reveal_type(Foo())  # revealed: Foo
```

## A descriptor in place of `__new__`

```py
class SomeCallable:
    def __call__(self, cls, x: int) -> "Foo":
        obj = object.__new__(cls)
        obj.x = x
        return obj

class Descriptor:
    def __get__(self, instance, owner) -> SomeCallable:

>         print(instance, owner)

Reviewer 2: Not much point in having this in an mdtest?
```suggestion
```

reveal_type(Foo(1, 2))  # revealed: Foo
```

## `__init__` present on a superclass

If the `__init__` method is defined on a superclass, we can still infer the signature of the
constructor from it.

```py
class Base:
    def __init__(self, x: int): ...

class Foo(Base): ...

reveal_type(Foo(1))  # revealed: Foo

# error: [missing-argument] "No argument provided for required parameter `x` of bound method `__init__`"
reveal_type(Foo())  # revealed: Foo
# error: [too-many-positional-arguments] "Too many positional arguments to bound method `__init__`: expected 1, got 2"
reveal_type(Foo(1, 2))  # revealed: Foo
```

## Conditional `__init__`

```py
def _(flag: bool) -> None:
    class Foo:
        if flag:
            def __init__(self, x: int): ...
        else:
            def __init__(self, x: int, y: int = 1): ...

    reveal_type(Foo(1))  # revealed: Foo
    # error: [invalid-argument-type] "Object of type `Literal["1"]` cannot be assigned to parameter 2 (`x`) of bound method `__init__`; expected type `int`"
    reveal_type(Foo("1"))  # revealed: Foo
    # error: [missing-argument] "No argument provided for required parameter `x` of bound method `__init__`"

>     reveal_type(Foo())  # revealed: Foo

Reviewer 2: Same as above, test with two int args provided

# Constructor

When classes are instantiated, Python calls the meta-class `__call__` method, which can either be
customized by the user or `type.__call__` is used.

The latter calls the `__new__` method of the class, which is responsible for creating the instance
and then calls the `__init__` method on the resulting instance to initialize it with the same
arguments.

Both `__new__` and `__init__` are looked up using full descriptor protocol, but `__new__` is then
called as an implicit static, rather than bound method with `cls` passed as the first argument.

> `__init__` has no special handling, it is fetched as bound method and is called just like any other
> method.

Reviewer: Minor nitpick:
```suggestion
`__init__` has no special handling, it is fetched as bound method and is called just like any other
dunder method.
```

(Normal methods would also be looked up on the instance dictionary)

# Constructor

## No init or new

> Here we statically infer that there is no custom `__init__` or `__new__` method defined on the
> class, hence the call to the constructor is only valid with no arguments.

Reviewer: The reason that this works at all is that we fall back to calling `__init__` on `object`, correct?

`builtins.pyi`:
```pyi
class object:
    # …
    def __init__(self) -> None: ...
```

So maybe we can rephrase this a bit?
```suggestion
Every class has `object` in it's MRO, so if no `__init__` method is provided, we fall back to
`object.__init__`, which can only be called with zero arguments:
```

```py
class Foo: ...

reveal_type(Foo())  # revealed: Foo

# error: [too-many-positional-arguments] "Too many positional arguments to bound method `__init__`: expected 1, got 2"
Foo(1)
```

## `__init__` present on the class itself

If the class has an `__init__` method, we can infer the signature of the constructor from it.

```py
class Foo:
    def __init__(self, x: int): ...

reveal_type(Foo(1))  # revealed: Foo

# error: [missing-argument] "No argument provided for required parameter `x` of bound method `__init__`"
Foo()
# error: [too-many-positional-arguments] "Too many positional arguments to bound method `__init__`: expected 2, got 3"

> Foo(1, 2)

Reviewer: Could we also add `reveal_type(…)`s here, to see what type we infer for those incorrect `Foo(…)` calls? I think ideally, we would still infer `Foo` in all cases.

Author: will do! and yes, that's exactly what I mentioned in my comment about `CallBinding` API's. I also think that we should emit diagnostic but still infer as if the call was correct
